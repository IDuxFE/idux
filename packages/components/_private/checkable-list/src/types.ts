/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { VirtualScrollToFn } from '@idux/cdk/scroll'
import type { ExtractInnerPropTypes, ExtractPublicPropTypes, MaybeArray, VKey } from '@idux/cdk/utils'
import type { DefineComponent, HTMLAttributes, PropType } from 'vue'

export interface CheckableListData {
  key?: VKey
  label?: string
  disabled?: boolean
  [key: string]: unknown
}

export interface CheckableListScroll {
  height?: string | number
  fullHeight?: boolean
}

export const checkableListProps = {
  dataSource: Array as PropType<CheckableListData[]>,
  checkable: {
    type: Boolean,
    default: true,
  },
  removable: {
    type: Boolean,
    default: false,
  },
  checked: Function as PropType<(item: CheckableListData) => boolean>,
  customAdditional: { type: Function as PropType<CheckableListCustomAdditional>, default: undefined },
  disabled: Function as PropType<(item: CheckableListData) => boolean>,
  getKey: Function as PropType<(item: CheckableListData) => VKey>,
  labelKey: String,
  virtual: {
    type: Boolean,
    default: false,
  },
  scroll: Object as PropType<CheckableListScroll>,
  onCheckChange: [Function, Array] as PropType<MaybeArray<(item: CheckableListData, checked: boolean) => void>>,
  onRemove: [Function, Array] as PropType<MaybeArray<(item: CheckableListData) => void>>,
  onScroll: [Function, Array] as PropType<MaybeArray<(evt: Event) => void>>,
  onScrolledChange: [Function, Array] as PropType<
    MaybeArray<(startIndex: number, endIndex: number, visibleData: unknown[]) => void>
  >,
  onScrolledBottom: [Function, Array] as PropType<MaybeArray<() => void>>,
} as const

export const checkableListItemProps = {
  checked: {
    type: Boolean,
    default: false,
  },
  checkable: {
    type: Boolean,
    default: true,
  },
  removable: {
    type: Boolean,
    default: false,
  },
  disabled: {
    type: Boolean,
    default: false,
  },
  label: String,
  value: {
    type: [String, Number, Symbol] as PropType<VKey>,
    required: true,
  },
  onCheckChange: [Function, Array] as PropType<MaybeArray<(checked: boolean) => void>>,
  onRemove: [Function, Array] as PropType<MaybeArray<() => void>>,
} as const

export interface CheckableListApi {
  scrollTo: VirtualScrollToFn
}

export type CheckableListProps = ExtractInnerPropTypes<typeof checkableListProps>
export type CheckableListItemProps = ExtractInnerPropTypes<typeof checkableListItemProps>
export type CheckableListPublicProps = ExtractPublicPropTypes<typeof checkableListProps>
export type CheckableListComponent = DefineComponent<
  Omit<HTMLAttributes, keyof CheckableListPublicProps> & CheckableListPublicProps,
  CheckableListApi
>
export type CheckableListInstance = InstanceType<DefineComponent<CheckableListProps, CheckableListApi>>

export type CheckableListCustomAdditional = (options: {
  data: CheckableListData
  index: number
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
}) => Record<string, any> | undefined
