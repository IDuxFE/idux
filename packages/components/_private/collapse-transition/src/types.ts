/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { ExtractInnerPropTypes, ExtractPublicPropTypes } from '@idux/cdk/utils'
import type { DefineComponent, HTMLAttributes } from 'vue'

import { IxPropTypes } from '@idux/cdk/utils'

export type CollapseTransitionMode = 'height' | 'width'

export const collapseTransitionProps = {
  appear: IxPropTypes.bool.def(false),
  name: IxPropTypes.string,
  mode: IxPropTypes.oneOf<CollapseTransitionMode>(['height', 'width']).def('height'),
  onAfterEnter: IxPropTypes.emit(),
  onAfterLeave: IxPropTypes.emit(),
}

export type CollapseTransitionProps = ExtractInnerPropTypes<typeof collapseTransitionProps>
export type CollapseTransitionPublicProps = ExtractPublicPropTypes<typeof collapseTransitionProps>
export type CollapseTransitionComponent = DefineComponent<
  Omit<HTMLAttributes, keyof CollapseTransitionPublicProps> & CollapseTransitionPublicProps
>
export type CollapseTransitionInstance = InstanceType<DefineComponent<CollapseTransitionProps>>
