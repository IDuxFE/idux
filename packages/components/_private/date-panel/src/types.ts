/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { ExtractInnerPropTypes, ExtractPublicPropTypes, MaybeArray } from '@idux/cdk/utils'
import type { DefineComponent, HTMLAttributes, PropType } from 'vue'

export const datePanelProps = {
  cellTooltip: Function as PropType<(cell: { value: Date; disabled: boolean }) => string | void>,
  disabledDate: Function as PropType<(date: Date) => boolean>,
  type: {
    type: String as PropType<DatePanelType>,
    default: 'date',
  },
  value: [Date, Array] as PropType<Date | (Date | undefined)[]>,
  activeDate: Date,
  visible: {
    type: Boolean,
    default: undefined,
  },
  isSelecting: {
    type: Boolean,
    default: undefined,
  },

  onCellClick: [Function, Array] as PropType<MaybeArray<(date: Date) => void>>,
  onCellMouseenter: [Function, Array] as PropType<MaybeArray<(date: Date) => void>>,
  'onUpdate:activeDate': [Function, Array] as PropType<MaybeArray<(date: Date) => void>>,
}

export type DatePanelProps = ExtractInnerPropTypes<typeof datePanelProps>
export type DatePanelPublicProps = ExtractPublicPropTypes<typeof datePanelProps>
export type DatePanelComponent = DefineComponent<
  Omit<HTMLAttributes, keyof DatePanelPublicProps> & DatePanelPublicProps
>
export type DatePanelInstance = InstanceType<DefineComponent<DatePanelProps>>

export type DatePanelType = 'date' | 'week' | 'month' | 'quarter' | 'year'

// private
export const panelRowProps = {
  rowIndex: {
    type: Number,
    required: true,
  },
}

export const panelCellProps = {
  rowIndex: {
    type: Number,
    required: true,
  },
  cellIndex: {
    type: Number,
    required: true,
  },
}
