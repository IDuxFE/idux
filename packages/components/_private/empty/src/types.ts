/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { EmptyProps as IxEmptyProps } from '@idux/components/empty'

export interface EmptyProps {
  empty?: string | IxEmptyProps
}
