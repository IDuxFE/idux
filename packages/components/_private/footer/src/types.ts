/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { ExtractInnerPropTypes, ExtractPublicPropTypes, VKey } from '@idux/cdk/utils'
import type { ButtonProps } from '@idux/components/button'
import type { DefineComponent, HTMLAttributes, PropType, VNode } from 'vue'

export interface FooterButtonProps extends ButtonProps {
  key?: VKey
  text?: string | VNode
  onClick?: (evt: Event) => void
}

export const footerProps = {
  cancel: Function as PropType<(evt?: Event | unknown) => Promise<void> | void>,
  cancelButton: Object as PropType<ButtonProps>,
  cancelLoading: Boolean,
  cancelText: String,
  cancelVisible: {
    type: Boolean,
    default: true,
  },
  footer: [Boolean, Array, Object] as PropType<boolean | FooterButtonProps[] | VNode>,
  ok: Function as PropType<(evt?: Event | unknown) => Promise<void> | void>,
  okButton: Object as PropType<ButtonProps>,
  okLoading: Boolean,
  okText: String,
} as const

export type FooterProps = ExtractInnerPropTypes<typeof footerProps>
export type FooterPublicProps = ExtractPublicPropTypes<typeof footerProps>
export type FooterComponent = DefineComponent<Omit<HTMLAttributes, keyof FooterPublicProps> & FooterPublicProps>
export type FooterInstance = InstanceType<DefineComponent<FooterProps>>
