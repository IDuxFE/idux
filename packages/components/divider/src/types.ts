/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import { type DefineComponent, type HTMLAttributes } from 'vue'

import { type IxInnerPropTypes, IxPropTypes, type IxPublicPropTypes } from '@idux/cdk/utils'

export const dividerProps = {
  dashed: IxPropTypes.bool,
  label: IxPropTypes.string,
  labelPlacement: IxPropTypes.oneOf(['start', 'center', 'end'] as const),
  plain: IxPropTypes.bool,
  position: IxPropTypes.oneOf(['left', 'center', 'right'] as const),
  size: IxPropTypes.oneOf(['sm', 'md', 'lg'] as const),
  type: IxPropTypes.oneOf(['horizontal', 'vertical'] as const),
  vertical: IxPropTypes.bool,
}

export type DividerProps = IxInnerPropTypes<typeof dividerProps>
export type DividerPublicProps = IxPublicPropTypes<typeof dividerProps>
export type DividerComponent = DefineComponent<Omit<HTMLAttributes, keyof DividerPublicProps> & DividerPublicProps>
export type DividerInstance = InstanceType<DefineComponent<DividerProps>>
