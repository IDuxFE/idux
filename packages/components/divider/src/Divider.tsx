/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import { computed, defineComponent, normalizeClass } from 'vue'

import { Logger } from '@idux/cdk/utils'
import { useGlobalConfig } from '@idux/components/config'
import { covertLabelVNode } from '@idux/components/utils'

import { dividerProps } from './types'

export default defineComponent({
  name: 'IxDivider',
  props: dividerProps,
  setup(props, { slots }) {
    const common = useGlobalConfig('common')
    const mergedPrefixCls = computed(() => `${common.prefixCls}-divider`)
    const config = useGlobalConfig('divider')

    const classes = computed(() => {
      const {
        dashed = config.dashed,
        label,
        labelPlacement = config.labelPlacement,
        plain = config.plain,
        position,
        size = config.size,
        type,
        vertical,
      } = props

      __DEV__ &&
        position &&
        Logger.warn('components/divider', '`position` was deprecated, please use `labelPlacement` instead')
      __DEV__ && type && Logger.warn('components/divider', '`type` was deprecated, please use `vertical` instead')

      const withLabel = !!label || !!slots.default
      const prefixCls = mergedPrefixCls.value
      return normalizeClass({
        [`${prefixCls}`]: true,
        [`${prefixCls}-${size}`]: true,
        [`${prefixCls}-dashed`]: dashed,
        [`${prefixCls}-horizontal`]: !(type === 'vertical' || vertical),
        [`${prefixCls}-vertical`]: type === 'vertical' || vertical,
        [`${prefixCls}-plain`]: withLabel && plain,
        [`${prefixCls}-with-label`]: withLabel,
        [`${prefixCls}-with-label-${position || labelPlacement}`]: withLabel,
      })
    })

    return () => {
      const prefixCls = mergedPrefixCls.value
      const labelNode = covertLabelVNode(slots.default, props.label)
      return <div class={classes.value}>{labelNode && <span class={`${prefixCls}-label`}>{labelNode}</span>}</div>
    }
  },
})
