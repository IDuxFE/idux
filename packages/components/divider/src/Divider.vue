<template>
  <div :class="className">
    <span v-if="withText" :class="`${prefixCls}-divider-inner-text`">
      <slot />
    </span>
  </div>
</template>
<script lang="ts">
import type { ComputedRef } from 'vue'
import type { DividerConfig } from '@idux/components/config'
import type { DividerProps } from './types'

import { computed, defineComponent } from 'vue'
import { hasSlot } from '@idux/cdk/utils'
import { useGlobalConfig } from '@idux/components/config'
import { dividerProps } from './types'

export default defineComponent({
  name: 'IxDivider',
  props: dividerProps,
  setup(props, { slots }) {
    const { prefixCls } = useGlobalConfig('common')
    const dividerConfig = useGlobalConfig('divider')
    const withText = computed(() => hasSlot(slots))
    const className = useClassName(props, dividerConfig, withText)

    return { prefixCls, className, withText }
  },
})

function useClassName(props: DividerProps, config: DividerConfig, withText: ComputedRef<boolean>) {
  const { prefixCls } = useGlobalConfig('common')
  return computed(() => {
    const position = props.position || config.position
    const type = props.type || config.type
    const dashed = props.dashed || config.dashed
    const plain = props.plain || config.plain

    return [
      `${prefixCls}-divider`,
      `${prefixCls}-divider-${type}`,
      {
        [`${prefixCls}-divider-with-text`]: withText.value,
        [`${prefixCls}-divider-dashed`]: dashed,
        [`${prefixCls}-divider-plain`]: plain && withText.value,
        [`${prefixCls}-divider-with-text-${position}`]: type === 'horizontal' && withText.value,
      },
    ]
  })
}
</script>
