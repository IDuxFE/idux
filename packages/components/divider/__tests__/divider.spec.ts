import { MountingOptions, mount } from '@vue/test-utils'

import { renderWork } from '@tests'

import IxDivider from '../src/Divider'
import { DividerProps } from '../src/types'

describe('Divider', () => {
  const DividerMount = (options?: MountingOptions<Partial<DividerProps>>) => mount(IxDivider, { ...options })

  renderWork(IxDivider)

  test('dashed work', async () => {
    const wrapper = DividerMount({ props: { dashed: true } })

    expect(wrapper.classes()).toContain('ix-divider-dashed')

    await wrapper.setProps({ dashed: false })
    expect(wrapper.classes()).not.toContain('ix-divider-dashed')
  })

  test('label work', async () => {
    const label = 'label'
    const wrapper = DividerMount({ props: { label } })

    expect(wrapper.classes()).toContain('ix-divider-with-label')
    expect(wrapper.text()).toEqual(label)

    const label2 = `label2`
    await wrapper.setProps({ label: label2 })

    expect(wrapper.text()).toEqual(label2)
  })

  test('label slot work', async () => {
    const label = 'label'
    const labelSlot = 'label slot'
    const wrapper = DividerMount({ props: { label }, slots: { default: labelSlot } })

    expect(wrapper.text()).toEqual(labelSlot)
  })

  test('labelPlacement work', async () => {
    const wrapper = DividerMount({ props: { label: 'label', labelPlacement: 'start' } })

    expect(wrapper.classes()).toContain('ix-divider-with-label-start')

    await wrapper.setProps({ labelPlacement: 'end' })

    expect(wrapper.classes()).toContain('ix-divider-with-label-end')

    await wrapper.setProps({ label: undefined })

    expect(wrapper.classes()).not.toContain('ix-divider-with-label-end')
  })

  test('plain work', async () => {
    const wrapper = DividerMount({ props: { label: 'label', plain: true } })

    expect(wrapper.classes()).toContain('ix-divider-plain')

    await wrapper.setProps({ plain: false })

    expect(wrapper.classes()).not.toContain('ix-divider-plain')

    await wrapper.setProps({ label: undefined, plain: true })

    expect(wrapper.classes()).not.toContain('ix-divider-plain')
  })

  test('size work', async () => {
    const wrapper = DividerMount({ props: { size: 'sm' } })

    expect(wrapper.classes()).toContain('ix-divider-sm')

    await wrapper.setProps({ size: 'lg' })

    expect(wrapper.classes()).toContain('ix-divider-lg')
  })

  test('vertical work', async () => {
    const wrapper = DividerMount({ props: { vertical: true } })

    expect(wrapper.classes()).toContain('ix-divider-vertical')

    await wrapper.setProps({ vertical: false })
    expect(wrapper.classes()).toContain('ix-divider-horizontal')
  })
})
