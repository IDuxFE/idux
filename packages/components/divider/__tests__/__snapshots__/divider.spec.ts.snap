// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Divider render work 1`] = `
"<div class=\\"ix-divider ix-divider-md ix-divider-horizontal\\">
  <!---->
</div>"
`;
